File: src/tf_demo/src/dynamic_listener_tf.cpp
#include "geometry_msgs/PointStamped.h"
#include "ros/ros.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"

int main(int argc, char **argv) {
    setlocale(LC_ALL, "");
    ros::init(argc, argv, "dynamic_listener_tf");
    ros::NodeHandle nh;
    ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME, ros::console::levels::Debug);

    // 创建 TF 订阅对象并且监听
    tf2_ros::Buffer buffer;
    tf2_ros::TransformListener listener(buffer);

    ros::Rate rate(1);
    while (ros::ok()) {
        // 生成子级坐标系的座标点并配置
        geometry_msgs::PointStamped point_laser;
        point_laser.header.frame_id = "turtle1";
        point_laser.header.stamp = ros::Time();
        point_laser.point.x = 1;
        point_laser.point.y = 3;
        point_laser.point.z = 0;

        // 座标点转换
        try {
            geometry_msgs::PointStamped point_base;
            point_base = buffer.transform(point_laser, "world");
            ROS_DEBUG("\nTransformed position:\nx=%.2f\ny=%.2f\nCoordinate:\n%s\n",
                      point_base.point.x,
                      point_base.point.y,
                      point_base.header.frame_id.c_str());
        } catch (const tf2::TransformException &ex) {
            ROS_WARN("Failed to transform.");
        }
        rate.sleep();
    }
    return 0;
}

File: src/tf_demo/scripts/dynamic_listener_tf.py
import os

import rospy
import tf2_ros
from tf2_geometry_msgs import PointStamped


def main():
    rospy.init_node("dynamic_listener_tf")

    # 创建 TF 订阅对象
    buffer = tf2_ros.Buffer()
    listener = tf2_ros.TransformListener(buffer)

    rate = rospy.Rate(1)
    while not rospy.is_shutdown():
        point_laser = PointStamped()
        point_laser.header.frame_id = "turtle1"
        point_laser.header.stamp = rospy.Time.now()
        point_laser.point.x = 1
        point_laser.point.y = 3
        point_laser.point.z = 0.0

        try:
            point_base = buffer.transform(point_laser, "world", rospy.Duration(1))
            rospy.logdebug(
                "\nTransformed position:\nx=%.2f\ny=%.2f\n\Coordinate: %s\n",
                point_base.point.x,
                point_base.point.y,
                point_base.header.frame_id,
            )
        except:
            rospy.logwarn("Failed to transform.")
        rate.sleep()


if __name__ == "__main__":
    main()
